src/internal_modules/roc_sndio/target_sndfile/roc_sndio/sndfile_sink.cpp

```diff
@@ -0,0 +1,360 @@
+    roc_panic_if(!valid_);
+
+    roc_log(LogDebug, "sndfile sink: opening: driver=%s path=%s", driver, path);
+
+    if (buffer_.size() != 0 || sndfile_output_) {
+        roc_panic("sndfile sink: can't call open() more than once");
+    }
+
+    if (!open_(driver, path)) {
+        return false;
+    }
+
+    if (!setup_buffer_()) {
+        return false;
+    }
+
+    return true;
+}
+
+DeviceType SndfileSink::type() const {
+    return DeviceType_Sink;
+}
+
+DeviceState SndfileSink::state() const {
+    return DeviceState_Active;
+}
+
+void SndfileSink::pause() {
+    // no-op
+}
+
+bool SndfileSink::resume() {
+    return true;
+}
+
+bool SndfileSink::restart() {
+    return true;
+}
+
+audio::SampleSpec SndfileSink::sample_spec() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: sample_rate(): non-open output file or device");
+    }
+
+    if (sf_info_out_.channels == 1) {
+        return audio::SampleSpec(size_t(sf_info_out_.samplerate),
+                                 audio::ChanLayout_Surround, audio::ChanOrder_Smpte,
+                                 audio::ChanMask_Surround_Mono);
+    }
+
+    if (sf_info_out_.channels == 2) {
+        return audio::SampleSpec(size_t(sf_info_out_.samplerate),
+                                 audio::ChanLayout_Surround, audio::ChanOrder_Smpte,
+                                 audio::ChanMask_Surround_Stereo);
+    }
+
+    roc_panic("sndfile sink: unsupported channel count");
+}
+
+core::nanoseconds_t SndfileSink::latency() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: latency(): non-open output file or device");
+    }
+
+    return 0;
+}
+
+bool SndfileSink::has_latency() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: has_latency(): non-open output file or device");
+    }
+
+    return false;
+}
+
+bool SndfileSink::has_clock() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: has_clock(): non-open output file or device");
+    }
+
+    return !is_file_;
+}
+
+void SndfileSink::write(audio::Frame& frame) {
+    roc_panic_if(!valid_);
+    const audio::sample_t* frame_data = frame.samples();
+    size_t frame_size = frame.num_samples();
+    audio::sample_t* buffer_data = buffer_.data();
+    size_t buffer_pos = 0;
+
+    while (frame_size > 0) {
+        for (; buffer_pos < buffer_size_ && frame_size > 0; buffer_pos++) {
+            buffer_data[buffer_pos] = *frame_data;
+            frame_data++;
+            frame_size--;
+        }
+
+        if (buffer_pos > 0) {
+            if (sf_write_float(sndfile_output_, buffer_data, (sf_count_t)buffer_size_)
+                    == 0
+                && frame_size > 0) {
+                roc_panic(
+                    "sndfile sink: Unable to write entire frame to file. Reached eof");
```

Comment:
From my understanding sf_write_float has no error return as it only returns the number of items it was able to read, whether that be 0 or any other number. I can use sf_strerror(SNDFILE * file_) here if you like, but I think it will always return a SF_ERR_NO_ERROR = 0 message.

Reply:
Actually it seems it can raise quite a lot of errors: https://github.com/libsndfile/libsndfile/blob/e5ee50fbda1b9049a45fc65d06c34825feb4f237/src/sndfile.c#L2463

Most likely errors for us are failure to write to a file (e.g. disk is full, unmounted, etc).

You're right, seems we need sf_strerror().

I'd do something like this:

```
if (sf_write_float(...) != frame_size || sf_strerror(...) != 0) {
    // error
}
```

This way we'll catch all unusual situations. If everything is good, we should always write as many samples as requested, and error flag should not be set.